File: 08-安全/02-网络安全/WPA3-SAE/go/sae_group.go
// sae_group.go —— Dragonfly（WPA3-SAE）的群参数、KDF 与密码元素（RFC 7664 §2.2 / §3.2）
//
// 群：RFC 3526 §3 的 2048 位 MODP Group（id 14），p 是安全素数、G = 2、q = (p-1)/2。
// RFC 7664 把单向函数 H 与 KDF 留给上层协议，本仓库三个实现统一约定：
//
//	H   = SHA-256
//	KDF = HKDF-SHA256（Extract 用全零盐，Expand 用标签作为 info）
//
// 这两处「规范没规定」的地方正是跨语言实现最容易对不上的位置，所以三边都跑同一组向量。
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"math/big"
)

// RFC 3526 §3（2048-bit MODP Group, id 14）
const pHex = "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD129024E088A67CC74" +
	"020BBEA63B139B22514A08798E3404DDEF9519B3CD3A431B302B0A6DF25F1437" +
	"4FE1356D6D51C245E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED" +
	"EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3DC2007CB8A163BF05" +
	"98DA48361C55D39A69163FA8FD24CF5F83655D23DCA3AD961C62F356208552BB" +
	"9ED529077096966D670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B" +
	"E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9DE2BCBF695581718" +
	"3995497CEA956AE515D2261898FA051015728E5A8AACAA68FFFFFFFFFFFFFFFF"

const (
	pweLabel = "Dragonfly Hunting And Pecking"
	keyLabel = "Dragonfly Key Derivation"
	pBytes   = 256 // len(p) = 2048 位
)

var (
	p       = mustBig(pHex)
	one     = big.NewInt(1)
	two     = big.NewInt(2)
	pMinus1 = new(big.Int).Sub(p, one)
	q       = new(big.Int).Rsh(pMinus1, 1) // 子群阶 (p-1)/2
	g       = big.NewInt(2)
	expPeck = big.NewInt(2) // (p-1)/q，安全素数下恒为 2
)

func mustBig(hexStr string) *big.Int {
	v, ok := new(big.Int).SetString(hexStr, 16)
	if !ok {
		panic("非法的 16 进制大整数: " + hexStr)
	}
	return v
}

func qBytes() int { return (q.BitLen() + 7) / 8 }

func sha256Sum(data []byte) []byte {
	sum := sha256.Sum256(data)
	return sum[:]
}

func hmacSum(key, data []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}

// hkdfExpand：T(i) = HMAC(PRK, T(i-1) | info | i)，与 Python/C 版逐位一致
func hkdfExpand(prk, info []byte, n int) []byte {
	out := make([]byte, 0, n)
	var t []byte
	for ctr := byte(1); len(out) < n; ctr++ {
		mac := hmac.New(sha256.New, prk)
		mac.Write(t)
		mac.Write(info)
		mac.Write([]byte{ctr})
		t = mac.Sum(nil)
		out = append(out, t...)
	}
	return out[:n]
}

// kdf：HKDF-SHA256，Extract 盐为全零，Expand 用标签作 info
func kdf(label string, ikm []byte, nBytes int) *big.Int {
	prk := hmacSum(make([]byte, sha256.Size), ikm)
	return new(big.Int).SetBytes(hkdfExpand(prk, []byte(label), nBytes))
}

// ------------------------------------------------------------ 群运算（FFC）

// scalarOp：Z = Y^x mod p
func scalarOp(x, y *big.Int) *big.Int { return new(big.Int).Exp(y, x, p) }

// elementOp：Z = X·Y mod p
func elementOp(x, y *big.Int) *big.Int { return new(big.Int).Mod(new(big.Int).Mul(x, y), p) }

// inverse：R · R^-1 mod p = 1（p 是素数，故必存在）
func inverse(r *big.Int) *big.Int { return new(big.Int).ModInverse(r, p) }

// isValidElement：1 < e < p-1 且 e^q mod p == 1（在阶为 q 的子群内）
func isValidElement(e *big.Int) bool {
	if e == nil || e.Cmp(one) <= 0 || e.Cmp(pMinus1) >= 0 {
		return false
	}
	return new(big.Int).Exp(e, q, p).Cmp(one) == 0
}

// inRange：1 < x < q（规范对 private / mask / 对端 scalar 的共同要求）
func inRange(x *big.Int) bool {
	return x != nil && x.Cmp(one) > 0 && x.Cmp(q) < 0
}

// ------------------------------------------------------------ 密码元素

// huntingAndPecking：base = H(max(A,B)|min(A,B)|password|counter) →
//
//	seed = KDF-(len(p)+64)(base) mod (p-1) + 1 → temp = seed^((p-1)/q) mod p
//
// 找到 PE 之后仍然继续跑到第 k 轮，掩盖真实迭代数（RFC 7664 §3.2 的抗侧信道要求）。
func huntingAndPecking(password, idA, idB []byte, k int) (*big.Int, int) {
	hi, lo := idA, idB
	if bytes.Compare(idA, idB) < 0 { // 身份用 max/min 排序，双方视角才能算出同一个 PE
		hi, lo = idB, idA
	}
	nBytes := (p.BitLen() + 64 + 7) / 8 // 多取 64 位，降低模约减的偏置
	buf := make([]byte, 0, len(hi)+len(lo)+len(password)+1)
	buf = append(buf, hi...)
	buf = append(buf, lo...)
	buf = append(buf, password...)
	var pe *big.Int
	for counter := 1; ; counter++ {
		base := append(buf, byte(counter)) // len(buf) 固定，append 不会越过 cap
		seed := new(big.Int).Mod(kdf(pweLabel, sha256Sum(base), nBytes), pMinus1)
		seed.Add(seed, one)
		temp := scalarOp(expPeck, seed)
		if temp.Cmp(one) > 0 && pe == nil {
			pe = temp
		}
		if pe != nil && counter >= k {
			return pe, counter
		}
	}
}

func derivePWE(password, idA, idB []byte, k int) *big.Int {
	pe, _ := huntingAndPecking(password, idA, idB, k)
	return pe
}

// ------------------------------------------------------------ 序列化与比较

func bytesToHex(b []byte) string { return hex.EncodeToString(b) }

// fillBytes 定宽大端序列化。变长拼接会让不同的 (标量, 元素) 组合撞出同一输入串，
// 所以这里是硬要求：值超出 width 时 FillBytes 会 panic，而不是悄悄截断。
func fillBytes(v *big.Int, width int) []byte {
	buf := make([]byte, width)
	v.FillBytes(buf)
	return buf
}

// bigDigest：定宽序列化后的 sha256 摘要 —— 跨语言比 2048 位常量太笨重
func bigDigest(v *big.Int, width int) string {
	return bytesToHex(sha256Sum(fillBytes(v, width)))
}

// bigEqHex：与 16 进制串比较（Text(16) 不带前导零）
func bigEqHex(v *big.Int, hexStr string) bool { return v.Text(16) == hexStr }

File: 08-安全/02-网络安全/WPA3-SAE/python/sae_group.py
"""
Dragonfly 的有限域（MODP）群运算与密码元素推导（RFC 7664 §2.2 / §3.2）

群：RFC 3526 的 2048 位 MODP Group（id 14），p = 2^2048 - 2^1984 - 1 + 2^64*{[2^1918 pi] + 124476}，
生成元 G = 2，子群阶 q = (p-1)/2（p 是安全素数）。

为什么这个 demo 用 MODP 而不是 ECC：WPA3-Personal 的 SAE 实际部署主要跑椭圆曲线群，
但 RFC 7664 同时定义了 ECC 与 MODP 两条路径，而 MODP 路径**只需要大整数模幂**，
不用写点运算/求平方根，能在同样篇幅里把「狩猎与啄食」「提交/确认」讲清楚。
两侧的协议结构完全一样，换群只换 §2.1/§2.2 的算子。

hunting-and-pecking 里有个容易忽略的点：RFC 7664 要求**至少**迭代 k 次，
目的是不泄漏「找到 PE 用了几轮」这个侧信道（见 §3.2 的说明）。
"""

from __future__ import annotations

import hashlib
import hmac

# RFC 3526 §3（2048-bit MODP Group，id 14）的素数十六进制原文
P_HEX = ("FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD129024E088A67CC74"
         "020BBEA63B139B22514A08798E3404DDEF9519B3CD3A431B302B0A6DF25F1437"
         "4FE1356D6D51C245E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED"
         "EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3DC2007CB8A163BF05"
         "98DA48361C55D39A69163FA8FD24CF5F83655D23DCA3AD961C62F356208552BB"
         "9ED529077096966D670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B"
         "E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9DE2BCBF695581718"
         "3995497CEA956AE515D2261898FA051015728E5A8AACAA68FFFFFFFFFFFFFFFF")

P = int(P_HEX, 16)
G = 2
Q = (P - 1) // 2
P_BYTES = (P.bit_length() + 7) // 8

PWE_LABEL = b"Dragonfly Hunting And Pecking"
KEY_LABEL = b"Dragonfly Key Derivation"


def h(data: bytes) -> bytes:
    """RFC 7664 把单向函数 H 留给具体协议；WPA3-SAE 用 SHA-256。"""
    return hashlib.sha256(data).digest()


# ------------------------------------------------------------
# 1. 群运算（FFC）
# ------------------------------------------------------------

def scalar_op(x: int, element: int) -> int:
    """Z = scalar-op(x, Y) = Y^x mod p"""
    return pow(element, x, P)


def element_op(a: int, b: int) -> int:
    """Z = element-op(X, Y) = X*Y mod p"""
    return (a * b) % P


def inverse(element: int) -> int:
    """R * inverse(R) mod p = 1"""
    return pow(element, P - 2, P)


def is_valid_element(element: int) -> bool:
    """RFC 7664 §2.2：1 < element < p-1，且 element^q mod p == 1（在阶为 q 的子群内）。"""
    if not 1 < element < P - 1:
        return False
    return pow(element, Q, P) == 1


# ------------------------------------------------------------
# 2. KDF（HKDF-SHA256 的 Expand，标签作为 info）
# ------------------------------------------------------------

def kdf(label: bytes, ikm: bytes, n_bits: int) -> int:
    """输出 n_bits 位的伪随机串，返回其整数值。

    RFC 7664 只说「KDF-n」，没规定具体构造（这正是跨实现必须对齐的第一件事）。
    这里用 HKDF-SHA256：PRK = HMAC(zero_salt, ikm)，再按 info=label 展开。
    """
    prk = hmac.new(b"\x00" * 32, ikm, hashlib.sha256).digest()
    out, prev, counter = b"", b"", 1
    n_bytes = (n_bits + 7) // 8
    while len(out) < n_bytes:
        prev = hmac.new(prk, prev + label + bytes([counter]), hashlib.sha256).digest()
        out += prev
        counter += 1
    return int.from_bytes(out[:n_bytes], "big")


# ------------------------------------------------------------
# 3. 密码元素（PWE）—— 狩猎与啄食
# ------------------------------------------------------------

def hunting_and_pecking(password: bytes, id_a: bytes, id_b: bytes, k: int = 40,
                        nonce: bytes = b"") -> tuple[int, int]:
    """返回 (PE, 实际迭代次数)。

    身份用 max/min 排序后拼接，这样双方各自视角算出的 PE 相同 ——
    这是 Dragonfly 能做成对等（无固定角色）交换的关键。
    """
    hi, lo = max(id_a, id_b), min(id_a, id_b)
    n_bits = P.bit_length() + 64          # 多取 64 位，降低模约减的偏置
    pe, iterations = None, 0
    counter = 1
    while True:
        base = h(hi + lo + password + nonce + bytes([counter]))
        seed = (kdf(PWE_LABEL, base, n_bits) % (P - 1)) + 1
        temp = scalar_op((P - 1) // Q, seed)      # 安全素数下就是 seed^2 mod p
        if temp > 1 and pe is None:
            pe = temp
        iterations += 1
        counter += 1
        if pe is not None and iterations >= k:
            # 找到之后仍然继续跑，直到凑满 k 次，掩盖真实迭代数（防侧信道）
            return pe, iterations


def derive_pwe(password: bytes, id_a: bytes, id_b: bytes, k: int = 40,
               nonce: bytes = b"") -> int:
    return hunting_and_pecking(password, id_a, id_b, k, nonce)[0]


# ------------------------------------------------------------
# 4. 素数性自检（Miller-Rabin，仅用于验证群参数）
# ------------------------------------------------------------

def is_probable_prime(n: int, rounds: int = 12) -> bool:
    if n < 2:
        return False
    for small in (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37):
        if n % small == 0:
            return n == small
    d, s = n - 1, 0
    while d % 2 == 0:
        d //= 2
        s += 1
    for a in (2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37):
        x = pow(a, d, n)
        if x in (1, n - 1):
            continue
        for _ in range(s - 1):
            x = x * x % n
            if x == n - 1:
                break
        else:
            return False
    return True
